Moving configs to separate repo

haugene opened the pull request:
As discussed in #1489 and that we have been planning for a while. We should move all the .ovpn config files to a separate repo to make a clearer distinction between config and code. That would reduce the amount of PRs and issues posted here and we could set up a codeowner-structure or otherwise engage more of the community to maintain configs.

Separating the configs like this also enables people to fork, edit and test new configs without getting their PRs merged. While many of our contributors are familiar with git, this would allow any user to fork the config repo and copy-paste a config from his/her provider and be up and running.

The PR is not done. But it might not be far from it. I'm wondering what you think about naming of the variables and what we'll call these "modes" of operation. I'm not sure that we'll forever say that scripts live here, configs live there, so there should be some flexibility. For now this will default to "business as usual" and you have to say "static" config to make it download it. So it's opt-in.

As a final solution I'm up for suggestions, but I'm thinking that the default mode should be "auto" where we check if there are config scripts in the container for the given provider. If so we assume that they should be used to fetch/configure the provider. If we can't find them then we download the zip and look there. And then we fail if we can't find it.

If someone wanted to run their own scripts, they could then say "static" and point to their own repo and their scripts would be run instead. Calling it "static" is therefore a bit misleading for that scenario.

Anyways. I must be off for tonight! To be continued...

ksurl commented:
how about checking if the static config already exists locally (already downloaded before container restart)?

also partial git download to target a subfolder. I think there was a web app mentioned that requires hosting to enable as a middleware for this functionality. maybe a plugin for later versions.

haugene commented:
Good point. Should probably have the option to avoid downloading on every restart. :+1: But if you've targeted a branch then I think many want to get the latest version of that branch, so it'd be a opt-in to not download every time I would say. Or maybe we can store a hash and check if there are newer commits :thinking: 

About partial downloads. It sounds nice, but it would have to be both super simple and reliable to maybe be worth it imho. In the end, what are we saving? Megabytes on the download?

ksurl commented:
I think after cleaning up a few providers, the folder has shrank a lot. may not be an issue (it's under 100mb now).

haugene commented:
The vpn-configs repo zip is 23MB now. That's a benefit of doing zip download instead of git clone as well. We only get one revision and not the complete history, plus .ovpn files are quite zip-friendly. Lots of duplication -> good compression.

ksurl commented:
`git clone --depth=1` would also allow efficient git pulls on restarts if the local folder is already detected.

fetzu commented:
Just as a few thoughts: 

1. Why not have the script download only the config files for the set provider instead of all the configs? Maybe this could be achieved through a partial git pull on said subfolder.

2. How do you plan to handle providers that have a "dynamic" script (ie: configs pulled through an API)? Should the "static" configs still be allowed in the repo or should the repo be reserved exclusively for providers which don't have a "dynamic" config option?

3. With the planned feature of WireGuard implantation, should we already consider this with this PR ? AFAIK WireGuard configs would mostly be dynamic because of the key exchange; which means that the script would only be called in case OpenVPN is used/set.

(Sorry in advance if my remarks are out of bounds/useless..!)

haugene commented:
All input is appreciated @fetzu. And you make good points. Here are my remarks on a per-item basis.

1. This comes down to simplicity. Partial pulls on sub-folders sounds great, but the priority should be robustness. I've maintained this project for so long that I know that anything that can fail, will fail at some point. So I'm not making any arguments against it, other than - why not download it all? It's 23 MB of files where everything you don't need gets deleted later. If this was a IoT project on mobile networks then 23 MB could matter, but we're talking torrenting here. 23 MB is negligible. But I'm all for it as long as it's simple, clean and elegant. Just want to note [for example this commit](https://github.com/haugene/docker-transmission-openvpn/commit/e81b8fc293a1a86ef6f19e1e49e009bfe7f3a0a0) where we had issues with git cloning on arm devices. This is in the node build of a frontend and not in the main container, but you get my point. 
2. As I mention in the original description "static" and "dynamic" might be misleading terms. "internal" and "external" might be better. I had a thought that we would still keep scripts in the main repo. So any openvpn provider where we have automated download and extraction of config files should still be managed here. That way we can re-use common parts of the scripts across providers. The idea was to only extract ".ovpn" files to a separate repo to avoid tons of lines of code and many commits just updating static information in text files. But with the proposed setup nothing would stop you from adding scripts in the external repo and it would be run as if it was internal. The externally downloaded provider folder is just hot swapped in place.
3. I think we should plan for the WireGuard feature as well, at least a bit. I'm not a fan of trying to think our way to an optimal setup and how this will be when we have WireGuard. Then it's better to code something now and iterate on it in the future. But discussing it a bit now and making some right key decisions so that it won't bite us in the ass later is good. Always a hard balance to strike. Again. Dynamic/static might not be the right distinction for WireGuard either. I still have no experience in using WireGuard - but I assume there are config files here as well? Or would it be easier to implement it as scripts?

As an overall note. I'm pretty sure that we will not get the perfect solution on the first try on any of the points. And I don't even think we should aim for it, because it's always about improving and there is no one size fits all here. I want a working system we can improve, but where the core assumptions and mechanisms are relatively correct. If we fetch the configs by curl, git or crawling a specific twitter-feed is not my primary concern. That's just a building block we can exchange later.

haugene commented:
Updated the PR now and moved the download of the external configs to a separate script. For now it still defaults to downloading a zip from GitHub but we can have multiple configurable options. I've also set MODIFY_CHOSEN_CONFIG to be true by default. I think that most of the provider configs should be able to be added without any modifications and then they should be automatically modified to match the project setup. For some providers there might be need for custom tweaking, but they can then disable it.

Any thoughts @pkishino?

And. When we come to the point of merging this. We have a bunch of open PRs now for updated configs. Any suggestion on the way forward? We could close all the PRs and refer to the config-repo. But we could also merge them all and tag a 3.5 version of the image, and then merge this PR where we delete all the configs. Then I could sync the config-repo from the 3.5 tag and we would get all the PR changes over into the config repo.

ksurl commented:
I would suggest merging existing pr for config update and set a cutoff for future pr to be directed to the new repo. Easier than chasing down all the open pr to open a new one on the new repo

pkishino commented:
I agree, I held off on mergind the other PR so we could all agree an this, but I think that a 3.5 with all updated configs will be best to move forward.
then a 4.0 would be the aim for the split

amosshapira commented on openvpn/fetch-external-configs.sh:
```diff
@@ -0,0 +1,42 @@
+  # Create a temporary file and download bundle to it
+  config_repo_temp_zip_file=$(mktemp)
+  echo "Downloading configs from ${GITHUB_CONFIG_BUNDLE_URL} into ${config_repo_temp_zip_file}"
+  curl -sSL -o "${config_repo_temp_zip_file}" "${GITHUB_CONFIG_BUNDLE_URL}"
```
missing check of curl success

haugene replied:
what are you thinking here? do a `if [ $? -ne 0 ]; then` something? Or postfix the line with `|| exit 1` maybe? Or we could do a `set -e` to fail on any error? But we source the script from start.sh so setting -e might then leak into that script?

amosshapira replied:
As much as I dislike `set -e`, I've grown used to our team's standard shell script top of:

```
set -o nounset
set -o errexit
set -o pipefail
```

It might be a matter of taste, but that's why I don't like sourcing one script from another in this type of code, it's a bit like Basic `GOSUB` if you remember that. (https://hwiegman.home.xs4all.nl/gw-man/GOSUB.html) - it makes every variable a global variable.

Does this script HAVE to be sourced (with bash `source`/`.`) in order to set any variables or change other bash internal state (e.g. `cd ...`), or can it be executed as a sub-process?

haugene replied:
Good question. I'm used to sourcing them in this project because the sub scripts usually set environment variables. But I don't think we would need that here. So then add that header and do the cleanup in a trap - the end?

I've gotten some bash experience over the years and I'm capable. But I'm not very well versed in best practices. So I'm happy to get some pointers :+1:

amosshapira replied:
I prefer passing parameters to sub scripts instead of having them rely on environment variables (even if it means passing the environment variable as a parameter). This makes the sub script call more explicit and readable as well as makes running the sub script independently easier (e.g. while developing, or if you find that you want to use it from another point in the code).

haugene replied:
I'm planning a bigger rewrite at some point. Been coding a bit more Python lately and I think that a lot of the code here could be done cleaner and easier with Python. To take this one step at a time I don't think I'll get away from the env vars any time soon. The project is too reliant on them already. So I'll let it be for this script as well.

But I've added
```
set -o nounset
set -o errexit
set -o pipefail
```
as suggested. and also added `errexit` and `pipefail` to the main script. I'm not sure if it can stay there. We'll have to see how much issues it creates. There are steps that I've seen failing in many logs where the container manages without. Like when people mount /dev/net/tun and don't set CREATE_TUN_DEVICE=false. Then we get a "file exists" type of error but it's fine.

It is probably better to fail fast and add ignores for those cases where a failure is accepted or tolerated in certain scenarios. I'm marking this as resolved for now.

If you have time for another review as it stands now I'd be grateful. But I think it's closing in on merging to `dev` at least. And planning to do some hardening there. And updating documentation etc.

amosshapira commented on openvpn/fetch-external-configs.sh:
```diff
@@ -0,0 +1,42 @@
+  # Create a temporary folder and extract configs there
+  config_repo_temp_dir=$(mktemp -d)
+  echo "Extracting configs to ${config_repo_temp_dir}"
+  unzip -q "${config_repo_temp_zip_file}" -d "${config_repo_temp_dir}"
```
missing check of unzip success

haugene replied:
I set
```
set -o nounset
set -o errexit
set -o pipefail
```
in the script as proposed. it now fails if the unzip command fails.

amosshapira commented on openvpn/fetch-external-configs.sh:
```diff
@@ -0,0 +1,42 @@
+  mv "${provider_configs}" "${VPN_PROVIDER_HOME}"
+
+  # Clean up temporary files
+  rm -rf "${config_repo_temp_zip_file}" "${config_repo_temp_dir}"
```
Move this to run in a `trap`?

haugene replied:
good idea

haugene replied:
:heavy_check_mark:

amosshapira commented on openvpn/fetch-external-configs.sh:
```diff
@@ -0,0 +1,42 @@
+
+  # Clean up temporary files
+  rm -rf "${config_repo_temp_zip_file}" "${config_repo_temp_dir}"
+
```
missing explicit `exit 0`

haugene replied:
:heavy_check_mark:

amosshapira commented on openvpn/start.sh:
```diff
@@ -59,19 +59,43 @@ fi
-  MODIFY_CHOSEN_CONFIG=yeah
+fi
+
+if [[ -z ${CHOSEN_OPENVPN_CONFIG} ]]; then
```
missing double quotes

haugene replied:
I know that openvpn/start.sh has a bit of syntactical issues. Have run shellcheck on it and I'll do a fixup later. Don't want to bundle it with this change in case I have to roll back.

Either way. For the `-z` empty checks it doesn't seem to mind about skipping the quotes. Maybe because if the string is empty there will be no word splitting? And if it's not empty it doesn't really matter if the string is split because it's not empty anyways?

amosshapira replied:
I stand corrected. The need to surround parameters for `test -z` must be an anachronism.

amosshapira commented on openvpn/start.sh:
```diff
@@ -59,19 +59,43 @@ fi
+    . /etc/openvpn/fetch-external-configs.sh
+  fi
+
+  if [[ -x $VPN_PROVIDER_HOME/configure-openvpn.sh ]]; then
```
missing double quotes

amosshapira commented on openvpn/start.sh:
```diff
@@ -59,19 +59,43 @@ fi
+  fi
 fi
 
 if [[ -z ${CHOSEN_OPENVPN_CONFIG} ]]; then
```
missing double quotes

haugene commented:
Phew. Had to do some rebase magic here. GitHub did not appreciate the commit removing all the ovpn configs and related files.
This PR was very slow to load and it was impossible to see the actual diff. I've now committed the *.ovpn (++) removal to a separate branch `feature/remove-ovpn-config-files` and created a PR for that in #1769.

This PR will then target that branch so that we can review and discuss the actual code changes.

haugene commented:
what do you think @pkishino? Getting ready to merge to dev and start a small rewrite of the documentation?

pkishino commented:
yeah, finally some cleanup here :)
I'll proably be quite busy in the next week or two but try to look in from time to time and see what I can do
